src/amo/components/LandingPage/index.js

```diff
@@ -37,61 +37,88 @@ const ICON_MAP = {
 
 export class LandingPageBase extends React.Component {
   static propTypes = {
-    apiAddonType: PropTypes.func.isRequired,
-    contentForType: PropTypes.func,
+    addonTypeOfResults: PropTypes.string,
     dispatch: PropTypes.func.isRequired,
+    errorHandler: PropTypes.object.isRequired,
     featuredAddons: PropTypes.array,
     highlyRatedAddons: PropTypes.array,
+    loading: PropTypes.boolean,
     popularAddons: PropTypes.array,
     i18n: PropTypes.object.isRequired,
     params: PropTypes.objectOf({
       visibleAddonType: PropTypes.string.isRequired,
     }).isRequired,
+    resultsLoaded: PropTypes.bool.isRequired,
   }
 
   static defaultProps = {
-    apiAddonType: getApiAddonType,
-    contentForType: null,
+    loading: false,
+    resultsLoaded: false,
   }
 
   componentWillMount() {
+    const { params } = this.props;
+    if (!apiAddonTypeIsValid(params.visibleAddonType)) {
+      log.warn(oneLine`Skipping componentWillMount() because visibleAddonType
+        is invalid: ${params.visibleAddonType}`);
+      return;
+    }
+
+    this.getLandingDataIfNeeded();
     this.setViewContextType();
   }
 
   componentDidUpdate() {
+    const { params } = this.props;
+    if (!apiAddonTypeIsValid(params.visibleAddonType)) {
+      log.warn(oneLine`Skipping componentDidUpdate() because visibleAddonType
+        is invalid: ${params.visibleAddonType}`);
+      return;
+    }
+
+    this.getLandingDataIfNeeded();
     this.setViewContextType();
   }
 
-  setViewContextType() {
-    const { apiAddonType, dispatch, params } = this.props;
-
-    // This error is handled properly in `render()`, so we just ignore it
-    // and log here. The reason for this is this component gets loaded for
-    // what should be a 404 (/not-a-page/) because of limitations in our
-    // current router. See:
-    // https://github.com/mozilla/addons-frontend/issues/2161
-    try {
-      const addonType = apiAddonType(params.visibleAddonType);
-      dispatch(setViewContext(addonType));
-    } catch (err) {
-      if (err instanceof AddonTypeNotFound) {
-        log.info('AddonTypeNotFound in setViewContextType()', err);
-      } else {
-        throw err;
-      }
+  getLandingDataIfNeeded() {
```

Comment:
I don't think that would handle the case where data was loaded on the server. In that case, `componentWillMount` should do nothing because all of the data has already been loaded.